### PR TITLE
Update wcEcoli from Python 3.8 to 3.11.3 and up-to-date pip libraries

### LINKED ISSUE #1311: Upgrade to Python 3.11
Python 3.11rc2 is available and is slated to become Python 3.11 in a few weeks.

Python 3.9, 3.10, and 3.11 have compelling performance improvements and just a few code compatibility changes.

They have some minor new features, most notably fine grained error locations in tracebacks and type hint improvements that could help with numpy.

- [x] #1310 
- [x] [Wiki page](https://github.com/CovertLab/wcEcoli/wiki/Python-changes-since-3.8-%28summary%29) to summarize relevant Python changes.
   * The `collections.*` aliases to `collections.abc.*` were removed. wcEcoli imports `collections.Mapping` and `collections.Sequence`.
   * See the Wiki page for a few potential issues and deprecations.
   * See the Wiki page for some new language and library features.
- [x] Read library release notes and evaluate for compatibility issues.
   * Cython needs updating to 0.29.29+ to get compatibility fixes for CPython 3.10 and 3.11, so probably go with 0.29.32 (the latest release version) for more bug fixes. (Wait on 3.0 alpha builds.) [@1fish2]
   * numpy & scipy [@ggsun]
   * matplotlib [@ggsun]
   * aesara [@ggsun]
   * ...
- [ ] Test on Python 3.9.13+ (or skip it).
- [ ] Test on Python 3.10.6+ (or skip it).
- [ ] Test on Python 3.11.
- [ ] Evaluate moving on from the "actively discouraged" `distutils`, perhaps to a `pyproject.toml` file. See the latest [Python Packaging User Guide](https://packaging.python.org/en/latest/) and [PyPA Projects](https://packaging.python.org/en/latest/key_projects/).
- [x] Strip out some Python 2 compatibility code if it doesn't cause merge headaches for anyone.

### LINKED ISSUE #1365: M2 Apple Silicon
I installed the model on a new MacBook Pro with an M2 chip and had to upgrade the following libraries (for python 3.8.13). I let NumPy and SciPy install their own copy and let Aesara find NumPy's copy.

numpy 1.20.0 --> 1.23.4
scipy 1.7.3 --> 1.9.3
numba 0.55.0 --> 0.56.3
aesara 2.0.8 --> 2.8.7

These match the library versions @nvivanco is running on her M1.

For the older version of numpy, the lapack_opt_info for numpy was not available (not linked). The upgraded numpy version meant I needed to upgrade scipy and numba for compatibility. Finally, I had to upgrade aesara because the older version was giving`AttributeError: module 'numpy.distutils.__config__' has no attribute 'blas_opt_info'`

@1fish2 let me know if you have any thoughts

### PR DESCRIPTION
**Pull request by 1fish2**
Fixes #1311
Fixes #1365

* Python improvements include faster startup, faster runtime, better error messages, Structural Pattern Matching, dict union operators, and type hints like `list[str]` and `str | int`. (This PR doesn't use the new features.)
  _Python 3.11 is between 10-60% faster than Python 3.10. On average, we measured a 1.25x speedup on the standard benchmark suite._ -- That's on top of the 3.9 and 3.10 speedups.
  **BUT** a 1-CPU Sherlock node ran Parca + 2 sim gens only 4% faster. Maybe most of the model's computation is already within numpy, scipy, numba, cython, and aesara.
* The updated libraries fix the security vulnerabilities noted by GitHub's Dependabot.

**SHERLOCK ENVIRONMENT MODULES:**
* I added `openssl/3.0.7` to the module `/home/groups/mcovert/modules/wcEcoli/python3` (for CPython 3.10+) and commented out `glpk/4.55-sherlock2` and `openblas/0.2.19`, leftover from Python 2.
  * **NOTE:** Anyone still running Python 2 will need to load those 2 modules.
* There's a case for updating to newer native libs: libpng, freetype, sqlite, libuuid, ncurses, gcc.

**TODO**
- [x] Update to `numba==0.57.0` + `llvmlite==0.40.0` when they're released, and re-test. That team has release candidates and they're fixing bugs, esp. on Windows.
- [x] Test on Sherlock.
- [x] Test on Intel Mac.
- [ ] Test on M1 Mac.
- [x] Test on M2 Mac.
- [x] Test the macOS 13.3+ version of Apple's Accelerate library vs. OpenBLAS. Supposedly it's now usable and many times faster, at least on M1 and M2.
- [ ] Test on a Linux desktop.
- [x] Ideally, test on Windows.
- [ ] Ideally, test on Google Cloud Compute Engine.
- [ ] Remove `PYTHONWARNINGS=default` from `ecoli-pull-request.sh` and `run-fireworks.sh` if it's too spammy.
- [ ] Rebuild the `wcEcoli` pyenv before merging into master. (This PR uses `wcEcoli-staging`.)

### CODE REVIEW

**Comment by 1fish2** [comment]
To @rjuenemann and anyone else interested, while waiting for the Numba release it'd be great to test its release candidate (https://pypi.org/project/numba/#history) and the rest of this Python 3.11 PR on Apple Silicon and other platforms.

Furthermore, in macOS 13.3, Apple's Accelerate library can now substitute for OpenBLAS and should be considerably faster than OpenBLAS on Apple Silicon. On my Intel Mac, using Accelerate ran Parca + 2 sim gens 6% faster, which is probably just measurement variability.

**To test with Accelerate:** Build one pyenv per `requirements.txt` in this branch (`python-3.11.3`) and a second pyenv with the following install-Numpy command to compile Numpy to use Accelerate:

1. Save this text in `~/.numpy-site.cfg` (commenting out any existing settings):
    ```
    [accelerate]
    libraries = Accelerate, vecLib
    ```
2. `pip install --no-cache-dir numpy==1.24.2 --no-binary numpy`

**Optional steps to examine/test a pyenv**
```
python runscripts/debug/summarize_environment.py  # did numpy link to Accelerate?
runscripts/debug/time_libraries.sh
wholecell/tests/utils/test_openblas_threads.py
pytest
```

**Compare their Parca + sim elapsed CPU run times**
```
pyenv local wcEcoli-3.11  # or whatever you named it
make clean compile  # clear caches and .pyc files
python runscripts/manual/runParca.py -c6 python3.11  # or pick a different number of CPU cores than -c6
python runscripts/manual/runSim.py -g2 python3.11

pyenv local wcEcoli-3.11-accelerate  # or whatever you named it
make clean compile
python runscripts/manual/runParca.py -c6 python3.11-accel
python runscripts/manual/runSim.py -g2 python3.11-accel
```

**Optionally compare outputs**

1. Parca output
```
runscripts/debug/comparePickles.py out/python3.11/kb out/python3.11-accel/kb
```
2. Sim output, but I don't expect them to match.
```
runscripts/debug/diff_simouts.py out/python3.11-2/wildtype_000000/000000/generation_000001/000000/simOut/ out/python3.11-accel/wildtype_000000/000000/generation_000001/000000/simOut/
```
3. Run some analysis plots and compare them visually.

Thanks, and go ahead and log results in this PR.

**Comment by rjuenemann** [comment]
Thank you for the update @1fish2 ! I'll test the 3.11 PR and Accelerate on my M2 and report back.

**Comment by rjuenemann** [comment]
Hi @1fish2 -

I tested this out on my M2 last night. The setup went relatively smoothly and I was able to install the requirements at their specified versions. I did encounter some warnings in the pytest:

wcEcoli3.11
```
=== warnings summary ===
wholecell/io/tablereader.py:2
  /Users/rjuene/dev/wcEcoli/wholecell/io/tablereader.py:2: DeprecationWarning: 'chunk' is deprecated and slated for removal in Python 3.13
    from chunk import Chunk

wholecell/tests/utils/test_openblas_threads.py::Test_openblas_threads::test_openblas
  /Users/rjuene/dev/wcEcoli/wholecell/tests/utils/test_openblas_threads.py:66: UserWarning: Didn't reproduce the expected dot product using 1 OpenBLAS thread, 0.01668380558411259 != 0.016683805584112754, so simulation results aren't portable.
    warnings.warn(f"Didn't reproduce the expected dot product using 1"
```

wcEcoli3.11-accel
```
=== warnings summary===
wholecell/io/tablereader.py:2
  /Users/rjuene/dev/wcEcoli/wholecell/io/tablereader.py:2: DeprecationWarning: 'chunk' is deprecated and slated for removal in Python 3.13
    from chunk import Chunk

wholecell/tests/utils/test_openblas_threads.py::Test_openblas_threads::test_openblas
  /Users/rjuene/dev/wcEcoli/wholecell/tests/utils/test_openblas_threads.py:66: UserWarning: Didn't reproduce the expected dot product using 1 OpenBLAS thread, 0.0166838055841127 != 0.016683805584112754, so simulation results aren't portable.
    warnings.warn(f"Didn't reproduce the expected dot product using 1"
```

Using Accelerate unfortunately did not lead to speed up for me. It does look like numpy linked to accelerate.

```
python runscripts/manual/runParca.py -c6 python3.11 -  Elapsed time 265.60 sec (0:04:25.595855); CPU 129.43 sec
python runscripts/manual/runSim.py -g2 python3.11 - Elapsed time 1283.02 sec (0:21:23.024364); CPU 1282.77 sec

python runscripts/manual/runParca.py -c6 python3.11-accel - Elapsed time 261.72 sec (0:0
[parcaCompare.txt](https://github.com/CovertLab/wcEcoli/files/11356001/parcaCompare.txt)
4:21.720864); CPU 124.16 sec
python runscripts/manual/runSim.py -g2 python3.11-accel Elapsed time 1250.76 sec (0:20:50.755607); CPU 1251.34 sec
```

The Parca outputs seemed to have quite a few differences (7230 lines). I’ve attached the comparison. [parcaCompare.txt](https://github.com/CovertLab/wcEcoli/files/11356002/parcaCompare.txt)

For runtime comparison on my laptop with the master branch and older Python version:
```
python runscripts/manual/runParca.py -c6 python3old - Elapsed time 278.50 sec (0:04:38.497969); CPU 138.29 sec
python runscripts/manual/runSim.py -g2 python3old - Elapsed time 1224.85 sec (0:20:24.846500); CPU 1224.67 sec
```

Let me know what you think. I can test on my Windows machine next.

**Comment by 1fish2** [comment]
@rjuenemann That's very helpful!

* Go ahead and try Windows. See [this issue](https://github.com/numba/numba/issues/8841#issuecomment-1502036638) if it fails to load `llvmlite.dll`.
* I'll copy the chunk library to handle the (sad but true) deprecation.
* ... and add a timer to `test_openblas_threads.py` for a quick measure of Numpy + BLAS speed without running the WCM.
* Maybe we need to install numpy with `--no-use-pep517` as described [in these emails](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/KC4X3WP372B6W642SYYQVVBKAWVFAUNN/) to get Accelerate speedup, but other discussions said `--no-use-pep517` won't work, and it's not clear what that switch does.
* Floating point repeatability might be intractable. Any change to evaluation order can change the rounding of results. The `test_openblas_threads.py` just computes one dot product, asking OpenBLAS to use different numbers of threads, and that suffices to get different results. I wonder if we can control the number of Accelerate threads (`BNNSFilterParameters`?) from Numpy.

**Comment by rjuenemann** [comment]
Hi @1fish2, 

I’m happy to try to install numpy with `--no-use-pep517` and accelerate if we want another data point. Adding the timer to `test_openblas_threads.py` sounds useful!

I’ve now tried out the Python 3.11 version on my Windows machine.

I didn’t run into an issue with `llvmlite.dll`. I did encounter an issue installing the new python version with pyenv - `No module named _lizma`
Following some stack overflow pages, I sudo apt installed `liblzma_dev`, then uninstalled and reinstalled 3.11.3 with pyenv and it seemed to fix the issue. 

Summarize_environment and time_libraries looked okay. I passed all pytests except for failing test_open_blas_threads. Test output attached. 
[windowsOpenBlasTest.txt](https://github.com/CovertLab/wcEcoli/files/11362263/windowsOpenBlasTest.txt)

Again, I'm not seeing a noticeable difference in runtimes:
```
python runscripts/manual/runParca.py -c6 python3.11 - Elapsed time 477.51 sec (0:07:57.508392); CPU 210.47 sec
python runscripts/manual/runSim.py -g2 python3.11 - Elapsed time 1982.94 sec (0:33:02.936445); CPU 1982.12 sec
```

Compared to old runtimes on this machine:
```
python runscripts/manual/runParca.py -c6 python3old -  Elapsed time 437.31 sec (0:07:17.307120); CPU 200.79 sec
python runscripts/manual/runSim.py -g2 python3old - Elapsed time 2036.66 sec (0:33:56.664326); CPU 2034.45 sec
```
Thanks!

**Comment by 1fish2** [comment]
numba 0.57.0 and llvmlite 0.40.0 were finally released. Now:

- [ ] Test with numba 0.57.0, llvmlite 0.40.0, also numpy 1.24.3 and other recent library updates:
   - [ ] Intel Mac (@1fish2)
   - [ ] M1/M2 Mac (@rjuenemann)
   - [ ] Windows (@rjuenemann)
   - [ ] Sherlock (@1fish2)
   - [ ] Linux (@ggsun ?)
- [ ] Update installation docs.
- [ ] Fork the `chunk` library.

Installation notes:
  * @rjuenemann was that `sudo apt install liblzma_dev` in WSL on Windows?
  * Experimenting on Intel Mac, `pip install --no-binary numpy numpy==1.24.2` builds a Numpy that links to macOS Accelerate even without a `.numpy-site.cfg` file. So Accelerate must now be the default.
  * So to make Numpy use OpenBLAS requires (1) `pip install numpy==1.24.3` (precompiled wheels) or (2) `pip install --no-binary numpy numpy==1.24.3` with a `.numpy-site.cfg` file configured for OpenBLAS. That setup took twice as long to compute dot products in `test_openblas_threads.py`.
  * The `--no-use-pep517` argument didn't have any obvious impact.

**Note:** Starting in numpy 1.24.3, [integer conversions check for overflow](https://numpy.org/doc/stable/release/1.24.0-notes.html#conversion-of-out-of-bound-python-integers), e.g. `np.array([3000], dtype=np.int8)` will give a DeprecationWarning. This could make for a very noisy PR build.

**Comment by rjuenemann** [comment]
@1fish2  - yes that was `sudo apt install liblzma_dev` on WSL (Ubuntu distribution) on Windows. 

Thanks for the update - I will retest with the new library updates.

**Comment by rjuenemann** [comment]
Hi @1fish2  

Retested with the new library updates. I didn’t encounter anything unexpected. Still having warnings with OpenBLAS thread test on M2 and failing on Windows.


**M2 installing numpy with Accelerate**

=== warnings summary ===
wholecell/io/tablereader.py:2
  /Users/rjuene/dev/wcEcoli/wholecell/io/tablereader.py:2: DeprecationWarning: 'chunk' is deprecated and slated for removal in Python 3.13
    from chunk import Chunk

wholecell/tests/utils/test_openblas_threads.py::Test_openblas_threads::test_openblas
  /Users/rjuene/dev/wcEcoli/wholecell/tests/utils/test_openblas_threads.py:87: UserWarning: Didn't reproduce the expected dot product using 1 OpenBLAS thread, 0.0166838055841127 != 0.016683805584112754, so simulation results aren't portable.
    warnings.warn(f"Didn't reproduce the expected dot product using 1"

ParCa: Elapsed time 263.65 sec (0:04:23.653752); CPU 125.52 sec
2-gen simulation: Elapsed time 1282.43 sec (0:21:22.429618); CPU 1282.44 sec


**M2 installing numpy with built-in OpenBLAS**

=== warnings summary ===
wholecell/io/tablereader.py:2
  /Users/rjuene/dev/wcEcoli/wholecell/io/tablereader.py:2: DeprecationWarning: 'chunk' is deprecated and slated for removal in Python 3.13
    from chunk import Chunk

wholecell/tests/utils/test_openblas_threads.py::Test_openblas_threads::test_openblas
  /Users/rjuene/dev/wcEcoli/wholecell/tests/utils/test_openblas_threads.py:87: UserWarning: Didn't reproduce the expected dot product using 1 OpenBLAS thread, 0.01668380558411259 != 0.016683805584112754, so simulation results aren't portable.
    warnings.warn(f"Didn't reproduce the expected dot product using 1"

ParCa: Elapsed time 290.86 sec (0:04:50.857963); CPU 130.23 sec

2-gen simulation: Elapsed time 1291.40 sec (0:21:31.399535); CPU 1291.17 sec


**Windows installing numpy with built-in OpenBlas**

OpenBLAS thread test still failing [testOpenBlasUpdate.txt](https://github.com/CovertLab/wcEcoli/files/11424333/testOpenBlasUpdate.txt)

ParCa: Elapsed time 447.08 sec (0:07:27.084221); CPU 206.07 sec

2-gen simulation: Elapsed time 2049.56 sec (0:34:09.559818); CPU 2048.94 sec

**Comment by 1fish2** [comment]
Thanks, @rjuenemann!

I posted to the numpy discussion group asking how to make Accelerate really accelerate. I forked the Chunk library, fixed up the docstrings, and will add it to the repo after writing a unit test; maybe later move it to a PyPI library.

Not reproducing the expected dot product means we won't get portable results and thus results in Jenkins builds could differ from those on our development computers. Bummer for debugging.

**Comment by 1fish2** [comment]
I'm wondering whether it's worth renaming `test_openblas_threads.py` since it actually tests whatever BLAS library Numpy is linked to...

**Comment by rjuenemann** [comment]
> I'm wondering whether it's worth renaming `test_openblas_threads.py` since it actually tests whatever BLAS library Numpy is linked to...

Might not be a bad idea if it's not too big of a hassle

**Comment by 1fish2** [comment]
It turns out to use the new BLAS/LAPACK interfaces via macOS Accelerate, numpy has to define the C macro `ACCELERATE_NEW_LAPACK` before including the Accelerate and vecLib headers. Without  that and maybe other numpy changes, I think it'll hit "[bugs that cause wrong output under easily reproducible conditions](https://numpy.org/devdocs/user/building.html#accelerated-blas-lapack-libraries)" and miss out on 64-bit APIs.

The numpy folks expect a numpy PR to enable that "soon."

We could clone the numpy repo and tinker in the meantime or just wait for the PR.

**Comment by rjuenemann** [comment]
> It turns out to use the new BLAS/LAPACK interfaces via macOS Accelerate, numpy has to define the C macro `ACCELERATE_NEW_LAPACK` before including the Accelerate and vecLib headers. Without that and maybe other numpy changes, I think it'll hit "[bugs that cause wrong output under easily reproducible conditions](https://numpy.org/devdocs/user/building.html#accelerated-blas-lapack-libraries)" and miss out on 64-bit APIs.
> 
> The numpy folks expect a numpy PR to enable that "soon."
> 
> We could clone the numpy repo and tinker in the meantime or just wait for the PR.

Interesting. So this could explain why we weren't seeing a substantial speedup using Accelerate?

**Comment by 1fish2** [comment]
It seems that [compiling with](https://developer.apple.com/documentation/accelerate/blas) `ACCELERATE_NEW_LAPACK` (+ `ACCELERATE_LAPACK_ILP64`?) is needed to access the new 64-bit libraries which use M1 hardware features like [matrix hardware](https://github.com/JuliaLang/julia/issues/42312).

There are lots of variations in [installing](https://gist.github.com/MarkDana/a9481b8134cf38a556cf23e1e815dafb) [numpy](https://stackoverflow.com/questions/70240506/why-is-numpy-native-on-m1-max-greatly-slower-than-on-old-intel-i5) (e.g. via conda), [build](https://github.com/numpy/numpy/issues/23667#issuecomment-1533345911) details, math operations (should we benchmark matrix multiply?), array size, floating point precision, hardware (M1/M2 -/Pro/Max), and [bugs](https://github.com/numpy/numpy/pull/15759), so I'm unsure about it.

**Comment by rjuenemann** [comment]
Thanks @1fish2! I don't have a strong opinion about a convenient time to merge it in - anytime should be fine on my end.

File: docs/create-pyenv.md [hunk 1]
```diff
@@ -215,20 +227,42 @@ virtualenv.
...
+      libraries = Accelerate, vecLib
+      ```
+
+   In late 2023, numpy binary wheels are expected to use Accelerate on macOS 1.14+.
+
+   **Recommendation:** On macOS 13.3+, use
+   `pip install numpy==1.24.3 --no-binary numpy`
+   to link numpy to the macOS Accelerate library. Otherwise use
+   `pip install numpy==1.24.3`
+   to use numpy's built-in copy of OpenBLAS.
```
Comment (rjuenemann) [inline docs/create-pyenv.md:246]: I'm a little confused on the wording in this section (starting with line 242) vs line 192 above - are we recommending Mac users to use Accelerate? Or still use OpenBLAS for now?

Comment (1fish2) [reply docs/create-pyenv.md:246]: Good catch! I'll remove this until the Numpy team makes it work properly on Accelerate.

**Pull request merged**